pkg/operator/bootstrap.go

```diff
@@ -22,10 +27,115 @@ import (
 // was specified.
 const randomSecretSize = 64
 
-func (h *Handler) Bootstrap() (*regopapi.ImageRegistry, error) {
-	crList := &regopapi.ImageRegistryList{
+func resourceName(namespace string) string {
+	if namespace == "default" {
+		return "docker-registry"
+	}
+	return "image-registry"
+}
+
+func newSpecFromDeploymentConfig(dc *appsapi.DeploymentConfig) (imageregistryapi.ImageRegistrySpec, error) {
+	fail := func(format string, a ...interface{}) (imageregistryapi.ImageRegistrySpec, error) {
+		format = "unable to create an image registry spec from the DeploymentConfig %s/%s: " + format
+		a = append([]interface{}{dc.Namespace, dc.Name}, a...)
+		return imageregistryapi.ImageRegistrySpec{}, fmt.Errorf(format, a...)
+	}
+
+	containers := dc.Spec.Template.Spec.Containers
+	if len(containers) != 1 {
+		return fail("got %d containers, want 1", len("containers"))
+	}
+
+	registry := containers[0]
+	if registry.Name != "registry" {
+		return fail("got container named %q, want %q", registry.Name, "registry")
+	}
+
+	hardcodedEnvs := map[string]string{
+		"REGISTRY_HTTP_ADDR": ":5000",
+		"REGISTRY_HTTP_NET":  "tcp",
+	}
+	envs := map[string]string{}
+	for _, env := range registry.Env {
+		if env.ValueFrom != nil {
+			// FIXME(dmage): we should support it.
+			return fail("envrioment variable %s: valueFrom is not supported", env.Name)
+		}
+		if value, ok := hardcodedEnvs[env.Name]; ok {
+			if env.Value != value {
+				return fail("envrioment variable %s: got %q, but the only supported value is %q", env.Value, value)
+			}
+		} else if strings.HasPrefix(env.Name, "REGISTRY_") {
+			envs[env.Name] = env.Value
+		} else {
+			return fail("unsupported enviroment variable %s", env.Name)
+		}
+	}
+
+	// TODO(dmage): registry.EnvFrom?
+
+	// TODO(dmage): get config from registry.VolumeMounts
+
+	if _, ok := envs["REGISTRY_OPENSHIFT_SERVER_ADDR"]; !ok {
+		envs["REGISTRY_OPENSHIFT_SERVER_ADDR"] = "must-be-set-to-parse-the-config-file"
+	}
+	for key, value := range envs {
+		if err := os.Setenv(key, value); err != nil {
+			return fail("unable to set environment variable %s: %v", key, err)
+		}
+	}
+	config, extraConfig, err := configuration.Parse(bytes.NewReader([]byte(`version: 0.1
+storage:
+  cache:
+    blobdescriptor: inmemory
+  filesystem:
+    rootdirectory: /registry
```

Comment:
We need a configuration even if the config isn't mounted into the container from a secret. In this case we'll have to have a (hardcoded?) copy of the default configuration.

Reply:
In particular this logic will change the log level from debug to info on old installations, which might be good.